File: leetcode/src/02597_the_number_of_beautiful_subsets/solution.go
package leetcode

func beautifulSubsets(nums []int, k int) int {
	return dfs(nums, k, 0, 0) - 1
}

func dfs(nums []int, k int, i int, mask int) int {
	if i == len(nums) {
		return 1
	}
	notTake := dfs(nums, k, i+1, mask)
	for j := 0; j < i; j++ {
		if mask&(1<<j) > 0 && abs(nums[j]-nums[i]) == k {
			return notTake
		}
	}
	take := dfs(nums, k, i+1, mask|(1<<i))
	return notTake + take
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

File: leetcode/src/02597_the_number_of_beautiful_subsets/solution.rs
use std::collections::HashMap;

struct Solution;

impl Solution {
    pub fn beautiful_subsets(nums: Vec<i32>, k: i32) -> i32 {
        let mut nums = nums;
        nums.sort_unstable();
        Self::dfs(&nums, k, 0, &mut HashMap::new()) - 1
    }

    fn dfs(nums: &[i32], k: i32, i: usize, freq: &mut HashMap<i32, usize>) -> i32 {
        if i == nums.len() {
            return 1;
        }
        let mut res = Self::dfs(nums, k, i + 1, freq);
        if *freq.get(&(nums[i] - k)).unwrap_or(&0) == 0 {
            *freq.entry(nums[i]).or_default() += 1;
            res += Self::dfs(nums, k, i + 1, freq);
            *freq.entry(nums[i]).or_default() -= 1;
        }
        res
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() {
        let nums = vec![2, 4, 6];
        let k = 2;
        assert_eq!(Solution::beautiful_subsets(nums, k), 4);

        let nums = vec![1];
        let k = 1;
        assert_eq!(Solution::beautiful_subsets(nums, k), 1);
    }
}
